stan/math/prim/mat/fun/ocl_gpu.hpp

```diff
@@ -0,0 +1,296 @@
+#include <stan/math/prim/mat/fun/kern_gpu/matrix_inverse.cl>  // NOLINT
+      ;                                                       // NOLINT
+  kernel_strings["matrix_multiply"] =
+#include <stan/math/prim/mat/fun/kern_gpu/matrix_multiply.cl>  // NOLINT
+      ;                                                        // NOLINT
+  kernel_strings["basic_matrix"] =
+#include <stan/math/prim/mat/fun/kern_gpu/basic_matrix.cl>  // NOLINT
+      ;                                                     // NOLINT
+
+  // Check if the kernels were already compiled
+  compiled_kernels["basic_matrix"] = false;
+  compiled_kernels["matrix_multiply"] = false;
+  compiled_kernels["timing"] = false;
+  compiled_kernels["matrix_inverse"] = false;
+  compiled_kernels["cholesky_decomposition"] = false;
+  compiled_kernels["check_gpu"] = false;
+}
+
+// TODO(Rok): select some other platform/device than 0
+// TODO(Rok): option to turn profiling OFF
+/**
+ * The class that represents the OpenCL runtime
+ *
+ * The class is used to find the OpenCL supported
+ * platforms/devices and initialize the OpenCL
+ * context/command queue.
+ *
+ */
+class ocl {
+ private:
+  std::string description_;
+  cl::Context oclContext_;
+  cl::CommandQueue oclQueue_;
+  cl::Platform oclPlatform_;
+  cl::Device oclDevice_;
+
+  size_t max_workgroup_size;
+
+  inline void init() {
+    try {
+      std::vector<cl::Platform> allPlatforms;
+      cl::Platform::get(&allPlatforms);
+      if (allPlatforms.size() == 0) {
+        domain_error("OpenCL Initialization", "[Platform]", "",
+                     "No OpenCL platforms found");
+      }
+      oclPlatform_ = allPlatforms[0];
+      std::vector<cl::Device> allDevices;
+      oclPlatform_.getDevices(DEVICE_FILTER, &allDevices);
+      // TODO(Steve): This should throw which platform
+      if (allDevices.size() == 0) {
+        domain_error("OpenCL Initialization", "[Device]", "",
+                     "No OpenCL devices found on the selected platform.");
+      }
+      oclDevice_ = allDevices[0];
+      description_ = "Device " + oclDevice_.getInfo<CL_DEVICE_NAME>()
+                     + " on the platform "
+                     + oclPlatform_.getInfo<CL_PLATFORM_NAME>();
+      allDevices[0].getInfo<size_t>(CL_DEVICE_MAX_WORK_GROUP_SIZE,
+                                    &max_workgroup_size);
+      oclContext_ = cl::Context(allDevices);
+      oclQueue_ = cl::CommandQueue(oclContext_, oclDevice_,
+                                   CL_QUEUE_PROFILING_ENABLE, NULL);
+      init_kernel_groups();
+      // Compile the dummy kernel used for timing purposes
+      cl::Program::Sources source(
+          1, std::make_pair(dummy_kernel.c_str(), dummy_kernel.size()));
+      cl::Program program_ = cl::Program(oclContext_, source);
+
+      try {
+        program_.build(allDevices);
+        kernels["dummy"] = cl::Kernel(program_, "dummy", NULL);
+        compiled_kernels["timing"] = true;
+      } catch (const cl::Error &e) {
```

Comment:
@SteveBronder, is this error not the same as the other errors in the build? (the outer try / catch)

Reply:
It is the same, we were throwing an error when the dummy kernel failed to compile and we get back some additional info. We can probably squish both of these together